starboard/shared/libiamf/iamf_decoder_utils.cc

```diff
@@ -0,0 +1,609 @@
+    *ptr = ByteSwap(*ptr);
+    pos_ += sizeof(uint32_t);
+    return true;
+  }
+
+  bool ReadLeb128(uint32_t* ptr) {
+    if (!HasBytes(sizeof(uint32_t)) || !ptr) {
+      return false;
+    }
+    int bytes_read = ReadLeb128Internal(buf_ + pos_, ptr);
+    if (bytes_read < 0) {
+      return false;
+    }
+    pos_ += bytes_read;
+    return true;
+  }
+
+  bool ReadString(std::string* str) {
+    int bytes_read = ReadStringInternal(buf_ + pos_, str);
+    if (bytes_read < 0) {
+      return false;
+    }
+    pos_ += bytes_read;
+    return true;
+  }
+
+  bool SkipBytes(size_t size) {
+    if (!HasBytes(size)) {
+      return false;
+    }
+    pos_ += size;
+    return true;
+  }
+
+  bool SkipLeb128() {
+    uint32_t val;
+    return ReadLeb128(&val);
+  }
+
+  bool SkipString() {
+    std::string str;
+    return ReadString(&str);
+  }
+
+  size_t size() const { return size_; }
+  size_t pos() const { return pos_; }
+  const uint8_t* buf() const { return buf_; }
+
+ private:
+  bool HasBytes(size_t size) const { return size + pos_ <= size_; }
+  inline uint32_t ByteSwap(uint32_t x) {
+#if defined(COMPILER_MSVC)
+    return _byteswap_ulong(x);
+#else
+    return __builtin_bswap32(x);
+#endif
+  }
+
+  // Decodes an Leb128 value and stores it in |value|. Returns the number of
+  // bytes read, capped to sizeof(uint32_t). Returns the number of bytes read,
+  // or -1 on error.
+  int ReadLeb128Internal(const uint8_t* buf, uint32_t* value) {
+    SB_DCHECK(buf);
+    SB_DCHECK(value);
+    *value = 0;
+    bool error = true;
+    size_t i = 0;
+    for (; i < sizeof(uint32_t); ++i) {
+      uint8_t byte = buf[i];
+      *value |= ((byte & 0x7f) << (i * 7));
+      if (!(byte & 0x80)) {
+        error = false;
+        break;
+      }
+    }
+
+    if (error) {
+      return -1;
+    }
+    return i + 1;
+  }
+
+  // Reads a c-string into |str|. Returns the number of bytes read, capped to
+  // 128 bytes, or -1 on error.
+  int ReadStringInternal(const uint8_t* buf, std::string* str) {
+    SB_DCHECK(buf);
+
+    int remaining_size = static_cast<int>(size_) - pos_;
+    if (remaining_size <= 0) {
+      return -1;
+    }
+
+    // The size of the string is capped to 128 bytes.
+    const int kMaxStringSize = 128;
+    int str_size = std::min(remaining_size, kMaxStringSize);
+    str->clear();
+
+    size_t bytes_read = 0;
+    while (bytes_read < str_size && buf[bytes_read] != '\0') {
+      bytes_read++;
+    }
+
+    if (bytes_read == str_size) {
+      if (buf[bytes_read - 1] != '\0') {
+        return -1;
+      }
+    } else {
+      if (buf[bytes_read] != '\0') {
+        return -1;
+      }
+    }
+
+    if (bytes_read > 0) {
+      str->resize(bytes_read);
+      std::memcpy(str->data(), reinterpret_cast<const char*>(buf), bytes_read);
+    }
+
+    // Account for null terminator byte.
+    return ++bytes_read;
+  }
+
+  int pos_ = 0;
+  const uint8_t* buf_;
+  const size_t size_ = 0;
```

Comment:
We don't have to initialize it to 0, since it's const, it will be forced by the compiler to initialize in ctor.